app/services/ProcessCsvService.scala

```diff
@@ -0,0 +1,273 @@
+import uk.gov.hmrc.services.validation.DataValidator
+import uk.gov.hmrc.services.validation.models._
+import utils.{ErrorResponseMessages, ValidationUtils}
+
+import scala.collection.mutable.ListBuffer
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success, Try}
+
+import javax.inject.{Inject, Singleton}
+
+@Singleton
+class ProcessCsvService @Inject()(auditEvents: AuditEvents,
+                                  dataGenerator: DataGenerator,
+                                  appConfig: ApplicationConfig,
+                                  ersConnector: ERSFileValidatorConnector
+                                 )(implicit executionContext: ExecutionContext,
+                                   actorSystem: ActorSystem) {
+
+  private val uploadCsvSizeLimit: Int = appConfig.uploadCsvSizeLimit
+  private[services] val ersSheetsClone: Map[String, SheetInfo] = ersSheets
+
+  def extractEntityData(response: HttpResponse): Source[ByteString, _] =
+    response match {
+      case HttpResponse(akka.http.scaladsl.model.StatusCodes.OK, _, entity, _) => entity.withSizeLimit(uploadCsvSizeLimit).dataBytes
+      case notOkResponse =>
+        Logger.error(
+          s"[ProcessCsvService][extractEntityData] Illegal response from Upscan: ${notOkResponse.status.intValue}, " +
+            s"body: ${notOkResponse.entity.dataBytes}")
+        Source.failed(
+          ERSFileProcessingException(
+            s"${ErrorResponseMessages.fileProcessingServiceFailedStream}",
+            s"${ErrorResponseMessages.fileProcessingServiceBulkEntity}"))
+    }
+
+  def extractBodyOfRequest: Source[HttpResponse, _] => Source[Either[Throwable, List[ByteString]], _] =
+    _.flatMapConcat(extractEntityData)
+      .via(CsvParsing.lineScanner())
+      .via(Flow.fromFunction(Right(_)))
+      .recover {
+        case e => Left(e)
+      }
+
+  def processFiles(callback: UpscanCsvFileData, source: String => Source[HttpResponse, _])(
+    implicit request: Request[_], hc: HeaderCarrier
+  ): List[Future[Either[Throwable, CsvFileContents]]] =
+    callback.callbackData map { successUpload =>
+
+      val sheetName = successUpload.name.replace(".csv", "")
+      val tryValidator: Either[Throwable, (DataValidator, SheetInfo)] = dataGenerator.getValidatorAndSheetInfo(sheetName, callback.schemeInfo)
+
+      tryValidator.fold(
+        throwable => Future.successful(Left(throwable)),
+        setValidator => {
+          val validator = setValidator._1
+          val sheetInfo = setValidator._2
+          val futureListOfErrors: Future[Seq[Either[Throwable, Seq[String]]]] = extractBodyOfRequest(source(successUpload.downloadUrl))
+            .via(eitherFromFunction(processRow(_, successUpload.name, callback.schemeInfo, validator, sheetInfo)))
+            .takeWhile(_.isRight, inclusive = true)
+            .runWith(Sink.seq[Either[Throwable, Seq[String]]])
+
+          futureListOfErrors.map { sequenceOfEithers =>
```

Comment:
I wonder here rather than calling `collectFirst` and doing a case match, if it would be more worthwhile taking the head of the sequence and folding? It would seem clearer to me, although that's subjective. Something like this: 

```suggestion
            futureListOfErrors.map { seq =>
            seq.head.fold(
              throwable => Left(throwable),
              _ => Right(CsvFileSubmissions(sheetName, seq.length, successUpload)))
          }
```

Reply:
while looking into this I realised there's a small edge case here: if the file is empty (0 bytes), in the version we already have there it will consider it 'a happy file' (and fail a bit farther down the line), whereas in the version you're suggesting it would be 'unhappy' (.head, or, more likely, .last, called on an empty sequence). we're having a quick look into what we should optimally be doing - will get back on this shortly.